modules/biobank/php/containercontroller.class.inc

```diff
@@ -0,0 +1,405 @@
+     * to provide the permissable Container Objects for the current User.
+     *
+     * @return \LORIS\Data\Provisioner
+     */
+    private function getDataProvisioner() : \LORIS\Data\Provisioner
+    {
+        $dao = new ContainerDAO($this->db);
+        if ($this->user->hasPermission('access_all_profiles') === false) {
+            $dao = $dao->filter(new \LORIS\Data\Filters\UserSiteMatch());
+        }
+        if ($this->user->hasPermission('biobank_container_view') === false) {
+            $dao = $dao->filter(new PrimaryContainerFilter(0));
+        }
+        if ($this->user->hasPermission('biobank_specimen_view') === false) {
+            $dao = $dao->filter(new PrimaryContainerFilter(1));
+        }
+        return $dao;
+    }
+
+    /**
+     * Checks User Permissions for creating or updating Container Objects.
+     *
+     * @param string $action User requested action.
+     *
+     * @return void
+     */
+    private function validatePermission(string $action) : void
+    {
+        switch($action) {
+        case 'GET':
+            if (!$this->user->hasPermission('biobank_container_view')) {
+                throw new \Forbidden('Container: View Permission Denied');
+            }
+        case 'PUT':
+            if (!$this->user->hasPermission('biobank_container_update')) {
+                throw new \Forbidden('Container: Update Permission Denied');
+            }
+        case 'POST':
+            if (!$this->user->hasPermission('biobank_container_create')) {
+                throw new \Forbidden('Container: Create Permission Denied');
+            }
+        }
+    }
+
+    /**
+     * Validates Container Object before being saved
+     *
+     * @param Container $container Container to be validated
+     *
+     * @return void
+     */
+    public function validateInstance(Container $container) : void
+    {
+        $this->validateBarcode($container);
+        $this->validateTypeId($container);
+        $this->validateTemperature($container);
+        $this->validateStatusId($container);
+        $this->validateCenterId($container);
+        $this->validateParentContainerId($container);
+        $this->validateCoordinate($container);
+    }
+
+    /**
+     * Validates Container Object Barcode
+     *
+     * @param Container $container Container to be checked.
+     *
+     * @throws BadRequest if the provided Container does not meet validation requirements
+     *
+     * @return void
+     */
+    private function validateBarcode($container) : void
+    {
+        $barcode = $container->getBarcode();
+        if (is_null($barcode)) {
+            throw new \BadRequest('Barcode value must be set');
+        }
+
+        $id = $container->getId();
+        if (!isset($id)) {
+            $containerList = $this->dao->selectInstances();
+            foreach ($containerList as $c) {
+                $b = $c->getBarcode();
+                if ($b === $barcode) {
```

Comment:
Can this be done with a unique key instead?

Reply:
I'm not sure what you mean.